fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java

```diff
@@ -624,11 +648,22 @@ private Query getFilterQuery(Set<ApiFilter> filters) {
         } finally {
             lock.readLock().unlock();
         }
-        return new SinglePagePagination<>(
+        page = new SinglePagePagination<>(
                 Collections.unmodifiableList(filteredDimRows.stream().collect(Collectors.toList())),
                 paginationParameters,
                 documentCount
         );
+        if (LUCENE_CACHE_SIZE > 0) {
+            try {
+                SinglePagePagination<DimensionRow> clone = SerializationUtils.clone(page);
+                synchronized (queryCache) {
+                    queryCache.put(new Pair<>(query, paginationParameters), clone);
+                }
```

Comment:
Actually JRE 1.8 has vastly improved synchronization handling.  Also, since this lock is only on put after long call to Lucene and of very brief duration, it should be very unlikely to cause contention.

However, if you know of a concurrent linked hash map that performs the required LRU mechanism, I would happily switch.